include/bio/map_io/header.hpp

```diff
@@ -0,0 +1,583 @@
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides the bio::map_io::tag_dictionary class and auxiliaries.
+ * \author Hannes Hauswedell <hannes.hauswedell AT decode.is>
+ */
+
+#pragma once
+
+#include <map>
+#include <regex>
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include <deque>
+#include <seqan3/std/ranges>
+#include <unordered_map>
+#include <vector>
+
+#include <seqan3/alphabet/concept.hpp>
+#include <seqan3/core/debug_stream.hpp>
+#include <seqan3/io/sam_file/detail/cigar.hpp>
+#include <seqan3/utility/views/single_pass_input.hpp>
+#include <seqan3/utility/views/type_reduce.hpp>
+
+#include <bio/exception.hpp>
+#include <bio/misc.hpp>
+
+namespace bio::map_io
+{
+
+/*!\brief The header used in mapping I/O.
+ * \ingroup map_io
+ * \details
+ *
+ * Each header line begins with the character `@` followed by one of the two-letter header record type codes
+ * defined in this section. In the header, each line is tab-delimited and, apart from `@CO` lines, each data field
+ * follows a format `TAG:VALUE` where TAG is a two-character string that defines the format and content of
+ * VALUE. Thus header lines match `/^@(HD|SQ|RG|PG)(\t[A-Za-z][A-Za-z0-9]:[ -~]+)+$/` or are comment lines staring
+ * with `@CO` followed by a tab and any character sequence.
+ * Within each (non-`@CO`) header line, no field tag may appear more than once and the order in which the fields
+ * appear is not significant.
+ *
+ * \sa https://samtools.github.io/hts-specs/SAMv1.pdf
+ */
+class header
+{
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    //!\brief Default construction.
+    header()                           = default; //!< Defaulted.
+    header(header const &)             = default; //!< Defaulted.
+    header(header &&)                  = default; //!< Defaulted.
+    ~header()                          = default; //!< Defaulted.
+    header & operator=(header const &) = default; //!< Defaulted.
+    header & operator=(header &&)      = default; //!< Defaulted.
+
+    /*!\brief Construct from a range of reference ids.
+     * \param[in] ref_ids The range over reference ids to redirect the pointer at.
+     */
+    template <typename ref_ids_type> // todo: restrict value type to be std::string_view constructible
+    header(ref_ids_type & ref_ids) : reference_names_given_on_construction{true}
+    {
+        for (auto & id : ref_ids)
+            reference_names.emplace_back(id);
+    }
+
+    /*!\brief Construct from a rvalue range of reference ids. Ids are copied over.
+     * \param[in] ref_ids The range over reference ids to own.
+     */
+    template <typename ref_ids_type> // todo: restrict value type to be std::string_view constructible
+    header(ref_ids_type && ref_ids) : reference_names_given_on_construction{true}
```

Comment:
This is a forwarding reference and not an rvalue reference. You need to restrict to that, too. 

Actually, it is even more tricky. The requirements are one of the following:
1. an rvalue container whose `value_t` is `std::string`; [this is hard to check, because we don't have a real container-concept];
2. any range (rvalue or not; container or not) whose `reference_t` is `std::string` or `std::string &&` (but not `std::string &`.

Reply:
As I said, I would remove this constructor and add `.set_rnames()` instead. But that has a similar problem, that rnames can take reference or ownership. I think this should work:
```c++
template <std::ranges::input_range rng_t>
    requires (std::constructible_from<std::string_view, std::ranges::range_reference_t<rng_t>>)
void set_rnames(rng_t && rng)
{
    // take references
    if constexpr (std::ranges::borrowed_range<rng_t> && std::ranges::borrowed_range<std::ranges::range_reference_t<rng_t>)
    {
        for (std::string_view const s : rng)
            reference_names.push_back(s);
    }
    // steal strings
    else if constexpr ((!std::ranges::borrowed_range<rng_t> && std::same_as<std::ranges::range_value_t<rng_t>, std::string>) ||
                       (detail::one_of<std::ranges::range_reference_t<rng_t>, std::string, std::string &&>))
    {
        for (std::string && s : rng)
        {
            owned_reference_names.push_back(std::move(s));
            reference_names.push_back(owned_reference_names.back());
        }
    }
    // copy strings
    else
    {
        for (std::string_view const s : rng)
        {
            owned_reference_names.push_back(static_cast<std::string>(s));
            reference_names.push_back(owned_reference_names.back());
        } 
    }
}
```